discord/http.py

```diff
@@ -260,123 +238,135 @@ async def request(
-                            is_global = data.get("global", False)
-                            if is_global:
-                                _log.warning(
-                                    (
-                                        "Global rate limit has been hit. Retrying in"
-                                        " %.2f seconds."
-                                    ),
-                                    retry_after,
+                            if remaining:
+                                remaining = int(remaining)
+                            if reset_after:
+                                reset_after = float(reset_after)
+
+                            bucket.set_metadata(remaining, reset_after)
+
+                            # the request was successful so just return the text/json
+                            if 300 > response.status >= 200:
+                                _log.debug("%s %s has received %s", method, url, data)
+                                return data
+
+                            # we are being rate limited
+                            if response.status == 429:
+                                if not response.headers.get("Via") or isinstance(
+                                    data, str
+                                ):
+                                    # Banned by Cloudflare more than likely.
+                                    raise HTTPException(response, data)
+
+                                fmt = (
+                                    "We are being rate limited. Retrying in %.2f seconds."
+                                    ' Handled under the bucket "%s"'
                                 )
-                                self._global_over.clear()
-
-                            await asyncio.sleep(retry_after)
-                            _log.debug("Done sleeping for the rate limit. Retrying...")
 
-                            # release the global lock now that the
-                            # global rate limit has passed
-                            if is_global:
-                                self._global_over.set()
-                                _log.debug("Global rate limit is now over.")
+                                # sleep a bit
+                                retry_after: float = data["retry_after"]
+                                _log.warning(fmt, retry_after, bucket_id)
+                                is_global: bool = data.get("global", False)
+
+                                if (
+                                    retry_after > self.maximum_rate_limit_time
+                                    and self.maximum_rate_limit_time != -1
+                                ):
+                                    raise HTTPException(
+                                        response,
+                                        f"Retrying rate limit would take longer than the maximum of {self.maximum_rate_limit_wait_time} seconds given",
```

Comment:
This one doesn't even exist lol
```suggestion
                                        f"Retrying rate limit would take longer than the maximum of {self.rate_limit_timeout} seconds given",
```